### PR TITLE
[Volvooncall] Updates to include Volvo V90 XC and to handle lock, heater and engine commands

### PR DESCRIPTION
**Pull request by kjoglum**
Updates to also enable actions for Volvo models being preclimatizationSupported rather than remoteHeaterSupported.
 
Also made changes to commands (heater/preclimatization and lock) as a "{}" attribute apparently is required included as part of the POST request sent to vocapi (as identified by bundle created for Homey).
 
Relate to ongoing discussion in OH forum:
https://community.openhab.org/t/solved-help-with-sending-command/7046/202
 
**NOTE:** Disregard changes/commit to 'VehicleStateDescriptionProvider.java' as the changes in my commit relate to changes needed to run the binding under my setup: OH 2.4.0 stable release. Later versions of volvooncall binding have implemented 'BaseDynamicStateDescriptionProvider' which apparently is not supported by 2.4.0 stable release.

### CODE REVIEW

File: bundles/org.openhab.binding.volvooncall/.classpath [hunk 1]
```diff
@@ -28,5 +28,14 @@
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
+	<classpathentry kind="lib" path="/Users/Kjoglum/.m2/repository/org/openhab/core/bundles/org.openhab.core.thing/2.5.0-SNAPSHOT/org.openhab.core.thing-2.5.0-SNAPSHOT.jar" sourcepath="/Users/Kjoglum/.m2/repository/org/openhab/core/bundles/org.openhab.core.thing/2.5.0-SNAPSHOT/org.openhab.core.thing-2.5.0-SNAPSHOT-sources.jar">
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/.classpath]: 1°/ This classpathentry countains local path definition.
2°/ This package is already included in the build so it does not have to be specifically included.

File: bundles/org.openhab.binding.volvooncall/README.md [hunk 2]
```diff
@@ -20,7 +20,7 @@ The binding has no configuration options itself, all configuration is done at 'T
 The 'VolvoOnCall API' bridge uses the owner's email address and password in order to access the VOC Remote API.
 This is the same email address and password as used in the VolvoOnCall smartphone app, that allows to remotely control your car(s).
 
-Once the bridge created, you will be able to launch discovery of the vehicles attached to it.
+Once the bridge created, you'll be able to launch discovery of the vehicles attached to it.
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/README.md]: Please, come back to previous way of writing. I think you need to rebase your local repository. This has been changed recently.

File: bundles/org.openhab.binding.volvooncall/pom.xml [hunk 3]
```diff
@@ -1,6 +1,7 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
 
-  <modelVersion>4.0.0</modelVersion>
+   <modelVersion>4.0.0</modelVersion>
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/pom.xml]: Extra space

File: bundles/org.openhab.binding.volvooncall/src/main/feature/feature.xml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<features name="org.openhab.binding.astro-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+<features name="org.openhab.binding.volvooncall-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
     <repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${project.version}/xml/features</repository>
 
-    <feature name="openhab-binding-volvooncall" description="Volvo On Call Binding" version="${project.version}">
+    <feature name="openhab-binding-volvooncall" description="volvooncall Binding" version="${project.version}">
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/feature/feature.xml]: Please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 5]
```diff
@@ -55,7 +55,7 @@ public void setThingHandler(@Nullable ThingHandler handler) {
 
     @RuleAction(label = "Volvo On Call : Close", description = "Closes the car")
     public void closeCarCommand() {
-        logger.debug("closeCarCommand called");
+        logger.warn("closeCarCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 6]
```diff
@@ -73,7 +73,7 @@ public static void closeCarCommand(@Nullable ThingActions actions) {
 
     @RuleAction(label = "Volvo On Call : Open", description = "Opens the car")
     public void openCarCommand() {
-        logger.debug("openCarCommand called");
+        logger.warn("openCarCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 7]
```diff
@@ -91,7 +91,7 @@ public static void openCarCommand(@Nullable ThingActions actions) {
 
     @RuleAction(label = "Volvo On Call : Start Engine", description = "Starts the engine")
     public void engineStartCommand(@ActionInput(name = "runtime", label = "Runtime") @Nullable Integer runtime) {
-        logger.debug("engineStartCommand called");
+        logger.warn("engineStartCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 8]
```diff
@@ -109,7 +109,7 @@ public static void engineStartCommand(@Nullable ThingActions actions, @Nullable
 
     @RuleAction(label = "Volvo On Call : Heater Start", description = "Starts car heater")
     public void heaterStartCommand() {
-        logger.debug("heaterStartCommand called");
+        logger.warn("heaterStartCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 9]
```diff
@@ -125,9 +125,27 @@ public static void heaterStartCommand(@Nullable ThingActions actions) {
...
+            ((VolvoOnCallActions) actions).preclimatizationStartCommand();
+        } else {
+            throw new IllegalArgumentException("Instance is not an VolvoThingActionsService class.");
+        }
+    }
+
     @RuleAction(label = "Volvo On Call : Heater Stop", description = "Stops car heater")
     public void heaterStopCommand() {
-        logger.debug("heaterStopCommand called");
+        logger.warn("heaterStopCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java [hunk 10]
```diff
@@ -143,10 +161,28 @@ public static void heaterStopCommand(@Nullable ThingActions actions) {
...
+        } else {
+            throw new IllegalArgumentException("Instance is not an VolvoThingActionsService class.");
+        }
+    }
+
     @RuleAction(label = "Volvo On Call : Honk-blink", description = "Activates the horn and or lights of the car")
     public void honkBlinkCommand(@ActionInput(name = "honk", label = "Honk") Boolean honk,
             @ActionInput(name = "blink", label = "Blink") Boolean blink) {
-        logger.debug("honkBlinkCommand called");
+        logger.warn("honkBlinkCommand called");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: This is not a warning, please revert.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 11]
```diff
@@ -414,6 +474,7 @@ public void actionStart(Integer runtime) {
 
                 try {
                     bridgeHandler.postURL(url, json);
+                    // logger.warn("Engine Start: {}", url + json);
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Please, remove

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleStateDescriptionProvider.java [hunk 12]
```diff
@@ -12,31 +12,57 @@
...
 
-    protected void unsetChannelTypeI18nLocalizationService(
-            final ChannelTypeI18nLocalizationService channelTypeI18nLocalizationService) {
-        this.channelTypeI18nLocalizationService = null;
+    @Deactivate
+    public void deactivate() {
+        this.channelOptionsMap.clear();
     }
-}
+}
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleStateDescriptionProvider.java]: You can not import a v2.4 file in a v2.5 build, or I do not get your point.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java [hunk 13]
```diff
@@ -67,9 +67,11 @@ public VolvoOnCallBridgeHandler(Bridge bridge) {
...
 
         httpHeader.put("cache-control", "no-cache");
-        httpHeader.put("X-OS-Type", "Android");
-        httpHeader.put("X-Originator-Type", "App");
-        httpHeader.put("Content-Type", JSON_CONTENT_TYPE);
+        httpHeader.put("content-type", JSON_CONTENT_TYPE);
+        httpHeader.put("x-device-id", "Device");
+        httpHeader.put("x-originator-type", "App");
+        httpHeader.put("x-os-type", "Android");
+        httpHeader.put("x-os-version", "22");
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Why these changes ?

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java [hunk 14]
```diff
@@ -159,6 +161,7 @@ public void run() {
     public void postURL(String URL, @Nullable String body) throws IOException, JsonSyntaxException {
         InputStream inputStream = body != null ? new ByteArrayInputStream(body.getBytes(StandardCharsets.UTF_8)) : null;
         String jsonString = HttpUtil.executeUrl("POST", URL, httpHeader, inputStream, null, REQUEST_TIMEOUT);
+        logger.warn("Post URL: {} Attributes {}", URL, httpHeader);
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: This is not a warning

**Comment by Hilbrand** [comment]
I would guess the file show up changed because the line endings have been changed from unix to windows. In which case that needs to be reverted.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java (hunk 13 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Not sure if required, although, included/corrected based on Volvo On Call bundle created for Athome Homey.

File: bundles/org.openhab.binding.volvooncall/README.md (hunk 2 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/feature/feature.xml (hunk 4 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/feature/feature.xml]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java (hunk 5 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java (hunk 7 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java (hunk 10 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java (hunk 11 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Corrected

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleStateDescriptionProvider.java (hunk 12 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleStateDescriptionProvider.java]: Corrected: New to git / pull requests etc, so I was hoping it was possible to skip some parts included in the pull request on the master side. Have now corrected pull request with the 2.5.0 file.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java (hunk 14 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Corrected.

**Comment by kjoglum** [comment]
Quite new to git / pull requests, and have no idea why the non-modified files are included in the PR. I have tried to rebase local repository as recommended, but they still show up as part of correction commits.

I have cloned the master repository, and have been using Eclipse IDE / Visual Studio Code for editing and mac terminal for git commands (running on Macbook).

Correction tips would be highly appreciated.

**Comment by wborn** [comment]
> Could it be the flags for the file have changed?

Changing the permissions on those files from 100644 to 100755 sets the executable permission which is wrong.

See also the recent https://github.com/openhab/openhab2-addons/pull/6249 where many such permissions were corrected.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 15]
```diff
@@ -380,23 +394,63 @@ private void actionHeater(String action, Boolean start) {
...
-                    bridgeHandler.postURL(SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command, null);
-                } catch (JsonSyntaxException | IOException e) {
-                    logger.warn("Exception occurred during execution: {}", e.getMessage(), e);
-                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                    if (command.contains("start")) {
+                        try {
+                            String address = SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command;
+                            String json = new String("{}");
+                            bridgeHandler.postURL(address, json);
+                        } catch (JsonSyntaxException | IOException e) {
```
[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Corrected in correction commit.

**Comment by kjoglum** [comment]
> > Could it be the flags for the file have changed?
> 
> Changing the permissions on those files from 100644 to 100755 sets the executable permission which is wrong.
> 
> See also the recent #6249 where many such permissions were corrected.

Running mac terminal / git version 2.23.0; have tried to chmod 644 the respective folders/files, but then, when trying to commit, I end up with "Permission denied".

Any detailed hints/guidelines?

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java [hunk 16]
```diff
@@ -159,6 +161,7 @@ public void run() {
     public void postURL(String URL, @Nullable String body) throws IOException, JsonSyntaxException {
         InputStream inputStream = body != null ? new ByteArrayInputStream(body.getBytes(StandardCharsets.UTF_8)) : null;
         String jsonString = HttpUtil.executeUrl("POST", URL, httpHeader, inputStream, null, REQUEST_TIMEOUT);
+        logger.debug("Post URL: {} Attributes {}", URL, httpHeader);
```
Comment (kjoglum) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Added to debug: To be able to check the actual POST command with attributes.

File: bundles/org.openhab.binding.volvooncall/.classpath [hunk 17]
```diff
@@ -29,4 +29,4 @@
 		</attributes>
 	</classpathentry>
 	<classpathentry kind="output" path="target/classes"/>
-</classpath>
+</classpath>
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/.classpath]: Extra space

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java (hunk 13 above)
Comment (clinique) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Can you try without these changes. I remind that I reduced it as much as possible in earlier versions of the binding.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java (hunk 16 above)
Comment (clinique) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Ok for me.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 18]
```diff
@@ -381,8 +390,29 @@ private void actionHeater(String action, Boolean start) {
...
+                            String address = SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command;
+                            String json = new String("{}");
+                            bridgeHandler.postURL(address, json);
+                        } else if (command.contains("stop")) {
+                            String address = SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command;
+                            bridgeHandler.postURL(address, null);
+                        }
+                    } else if (action.contains(PRECLIMATIZATION)) {
+                        String command = start ? "start" : "stop";
+                        if (command.contains("start")) {
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Simple remove the if and else construction and replace:
```
bridgeHandler.postURL(address, start? "{}" : null);

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java (hunk 13 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Have tried with and without the additions. Preclimatization works with both, however, lock/unlock seems to work only with the additions (for me at least).

Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java]: Have tried both with and without additions. Preclimatization works for both, however, lock function seem to work only with the additions (at least for me).

File: bundles/org.openhab.binding.volvooncall/.classpath (hunk 17 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/.classpath]: Corrected.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 19]
```diff
@@ -362,7 +364,14 @@ private void actionOpenClose(String action, OnOffType controlState) {
             if (activeOptions.containsKey(action)) {
                 if (vehicleStatus.carLocked != controlState) {
                     try {
-                        bridgeHandler.postURL(SERVICE_URL + "vehicles/" + configuration.vin + "/" + action, null);
+                        if (action.contains("lock")) {
```
[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Good point, corrected/combined in new commit.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java (hunk 18 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Corrected in new commit.

**Comment by kjoglum** [comment]
Did finally manage to exclude flagged/unmodified files from the PR:

Tried rebase of branch against master with no luck. Performing chmod 644 did not work either.

Had to replace the files using git checkout origin/master.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 20]
```diff
@@ -362,7 +364,9 @@ private void actionOpenClose(String action, OnOffType controlState) {
             if (activeOptions.containsKey(action)) {
                 if (vehicleStatus.carLocked != controlState) {
                     try {
-                        bridgeHandler.postURL(SERVICE_URL + "vehicles/" + configuration.vin + "/" + action, null);
+                        String address = SERVICE_URL + "vehicles/" + configuration.vin + "/" + action;
+                        String json = new String("{}");
+                        bridgeHandler.postURL(address, json);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Just use `"{}"` the `new String` is redundant and not used this way in java.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 21]
```diff
@@ -381,8 +385,15 @@ private void actionHeater(String action, Boolean start) {
         if (bridgeHandler != null) {
             if (activeOptions.containsKey(action)) {
                 try {
-                    String command = start ? "start" : "stop";
-                    bridgeHandler.postURL(SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command, null);
+                    if (action.contains(REMOTE_HEATER)) {
+                        String command = start? "start" : "stop";
+                        String address = SERVICE_URL + "vehicles/" + configuration.vin + "/heater/" + command;
+                        bridgeHandler.postURL(address, start? "{}" : null);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Space missing between  `start ?`.
Also missing below.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java (hunk 20 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Corrected with new commit.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java (hunk 21 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java]: Corrected with new commit.

File: bundles/org.openhab.binding.volvooncall/README.md [hunk 22]
```diff
@@ -99,13 +99,13 @@ Multiple actions are supported by this binding. In classic rules these are acces
 Example
 
 ```
- val actions = getActions("volvooncall","volvooncall:vehicle:bridgeId:thingId")
+ val actions = getActions("volvooncall","volvooncall:vehicle:myVinNumber")
  if(null === actions) {
         logInfo("actions", "Actions not found, check thing ID")
         return
- } else {
+    } else {
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/README.md]: Please review formatting.

File: bundles/org.openhab.binding.volvooncall/README.md [hunk 23]
```diff
@@ -99,13 +99,13 @@ Multiple actions are supported by this binding. In classic rules these are acces
...
         actions.openCarCommand()
- }
+  }
```
Comment (clinique) [inline bundles/org.openhab.binding.volvooncall/README.md]: Please review formatting.

File: bundles/org.openhab.binding.volvooncall/README.md [hunk 24]
```diff
@@ -99,13 +99,14 @@ Multiple actions are supported by this binding. In classic rules these are acces
 Example
 
 ```
- val actions = getActions("volvooncall","volvooncall:vehicle:bridgeId:thingId")
+ val actions = getActions("volvooncall","volvooncall:vehicle:myVinNumber")
```
[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: I do not understand how this can be wrong, as `getActions` is based on your version of the binding, and is the only command possible to use to trigger/send commands from script/rules (tested by far, i.e. proposed `actions.get` does not generate anything).

My only modification is the input String in `getActions`, which was needed to get commands through.

[the remark this answers is not in the document]
Comment (Hilbrand) [reply bundles/org.openhab.binding.volvooncall/README.md]: It's not wrong. It's technically possible to create a Thing with that UID. However, the official convention is to use `bridgeId:thingId`. But in this example it would probably be better to use the id of the other examples: `volvooncall:vehicle:glh:XC60`.

[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Well, the issue is that `bridgeid:thingId` i.e `glh:XC60` as referred to in the initial binding does not work with the commands/actions. I.e. getActions does not generate anything, and the commands will not be executed.

What does work is `volvooncall:vehicle:"VIN of the vehicle associated with this Thing"`

[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Meaning, `getActions("volvooncall", "volvooncall:vehicle:"VIN of the vehicle associated with this Thing").sendCommand ()` is the only setup in script/rule which works for me.

[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Also supported by other user testing same syntax, ref
https://community.openhab.org/t/solved-help-with-sending-command/7046/196

The command send is acknowledged and confirmed by the binding (although command is not executed correctly, something which this PR aim to fix with the {} attribute included amongst others.)

[the remark this answers is not in the document]
Comment (Hilbrand) [reply bundles/org.openhab.binding.volvooncall/README.md]: `volvooncall:vehicle:"VIN of the vehicle associated with this Thing"` only works because that is the definition of your thing and in your thing the `vin` value is set. The fact it works has nothing to do with your vin number. If you would set your thing to any other random name and use that name in the action it would still work. But depending on how you define a thing it can be either `volvooncall:vehicle:bridgeId:thingId` or `volvooncall:vehicle:thingId`. But the first is the actual preferred configuration.

[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Realised that as well. 

Although, in my case, the Thing, i.e. vehicle, was generated using autodiscovery, and then automatically became `volvooncall:vehicle:thingId` where thingID = VIN.

May need to have both alternatives presented in README, i.e. autodiscovered vs scriptbased Thing?

[the remark this answers is not in the document]
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Corrected with new commit.

File: bundles/org.openhab.binding.volvooncall/README.md (hunk 22 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Corrected.

File: bundles/org.openhab.binding.volvooncall/README.md (hunk 23 above)
Comment (kjoglum) [reply bundles/org.openhab.binding.volvooncall/README.md]: Corrected.

**Comment by kjoglum** [comment]
> 1 final request. The signoff needs to be your full name, not your github name.

Sorry for all the adjustments required from initial PR, but now, hopefully, the PR is complete and could be merged successfully.

Appreciate the guidance!
